File: sdk/LocationFetchService.m
//
//  LocationFetchService.m
//  mage-ios-sdk
//
//

#import "LocationFetchService.h"
#import "MageSessionManager.h"
#import "DataConnectionUtilities.h"
#import "MAGE-Swift.h"

NSString * const kLocationFetchFrequencyKey = @"userFetchFrequency";

@interface LocationFetchService ()
    @property (nonatomic) NSTimeInterval interval;
    @property (nonatomic, strong) NSTimer* locationFetchTimer;
@end

@implementation LocationFetchService

+ (instancetype) singleton {
    static LocationFetchService *fetchService = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        fetchService = [[self alloc] init];
    });
    return fetchService;
}

- (id) init {
    if (self = [super init]) {
        NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
        
        _interval = [[defaults valueForKey:kLocationFetchFrequencyKey] doubleValue];
        
        [[NSUserDefaults standardUserDefaults] addObserver:self
                                                forKeyPath:kLocationFetchFrequencyKey
                                                   options:NSKeyValueObservingOptionNew
                                                   context:NULL];
    }
	
	return self;
}

- (void) start {
    [self stop];

    [self pullLocations];
}

- (void) scheduleTimer {
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        weakSelf.locationFetchTimer = [NSTimer scheduledTimerWithTimeInterval:weakSelf.interval target:weakSelf selector:@selector(onTimerFire) userInfo:nil repeats:NO];
    });
}

- (void) onTimerFire {
    NSLog(@"timer to pull locations fired");
    if (![[UserUtility singleton] isTokenExpired]) {
        [self pullLocations];
    }
}

- (void) pullLocations{
    if ([DataConnectionUtilities shouldFetchLocations]) {
        NSURLSessionDataTask *locationFetchTask = [Location operationToPullLocationsWithSuccess:^(NSURLSessionDataTask * _Nonnull task, id _Nullable response) {
            if (![[UserUtility singleton] isTokenExpired]) {
                NSLog(@"Scheduling the location fetch timer");
                [self scheduleTimer];
            }
        } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull error) {
            NSLog(@"Failed to pull locations, scheduling the timer again");
            [self scheduleTimer];
        }];
        
        NSLog(@"pulling locations");
        [[MageSessionManager sharedManager] addTask:locationFetchTask];
    } else {
        [self scheduleTimer];
    }
}

-(void) stop {
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        if ([weakSelf.locationFetchTimer isValid]) {
            NSLog(@"Stopping the location fetch timer");
            [weakSelf.locationFetchTimer invalidate];
            weakSelf.locationFetchTimer = nil;
        }
    });
}

- (void) observeValueForKeyPath:(NSString *)keyPath
                      ofObject:(id)object
                        change:(NSDictionary *)change
                       context:(void *)context {
    _interval = [[change objectForKey:NSKeyValueChangeNewKey] doubleValue];
    [self start];
}


@end

File: sdk/LocationFetchService.swift
//
//  LocationFetchService.m
//  mage-ios-sdk
//
//

import Foundation
import Persistence
import ServerDTO
import UseCaseFactory
import LocationFetch

public class LocationFetchService: NSObject {
    
    public static let singleton = LocationFetchService()
    public var started = false
    
    var interval: TimeInterval = Double(UserDefaults.standard.userFetchFrequency)
    var locationFetchTimer: Timer?
    
    private override init() {
        super.init()
        UserDefaults.standard.addObserver(self, forKeyPath: "userFetchFrequency", options: .new, context: nil)
    }
    
    deinit {
        UserDefaults.standard.removeObserver(self, forKeyPath: "userFetchFrequency")
    }
    
    public override func observeValue(forKeyPath keyPath: String?, of object: Any?, change: [NSKeyValueChangeKey : Any]?, context: UnsafeMutableRawPointer?) {
        guard let change = change else {
            return
        }
        
        if change[NSKeyValueChangeKey.newKey] as? Double == interval {
            // we were called but thev alue is the same, ignore it
            return
        }
        if let interval = change[NSKeyValueChangeKey.newKey] as? Double {
            self.interval = interval
            if started {
                start()
            }
        }
    }
    
    public func start() {
        stop()
        pullLocations()
        started = true
    }
    
    public func stop() {
        NSLog("Stopping the location fetch timer")
        DispatchQueue.main.async { [weak self] in
            if let timer = self?.locationFetchTimer, timer.isValid {
                timer.invalidate()
                self?.locationFetchTimer = nil
            }
        }
        self.started = false
    }
    
    func pullLocations() {
        if !DataConnectionUtilities.shouldFetchLocations() {
            scheduleTimer()
            return
        }
        
        Task { [weak self] in
            defer {
                self?.scheduleTimer()
            }
            guard let currentEvent = Server.currentEventId() else {
                return
            }
            do {
                let eventID = EventID(currentEvent)
                let useCase = try await DependencyContainer.shared.useCaseFactory.resolve(
                    .EventLocationFetchUseCase
                )
                try await useCase
                    .execute(
                        eventID: eventID,
                        currentUserID: UserDefaults.standard
                            .currentUserId
                    )
                LocationFetchPackage.logger.info("Successfully fetched locations")
            } catch {
                LocationFetchPackage.logger.error("Failed to fetch locations: \(error.localizedDescription)")
            }
        }
    }
    
    func scheduleTimer() {
        if UserUtility.singleton.isTokenExpired {
            return;
        }
        if let locationFetchTimer = locationFetchTimer, locationFetchTimer.isValid {
            locationFetchTimer.invalidate()
            self.locationFetchTimer = nil
        }
        DispatchQueue.main.async { [weak self] in
            guard let fetchService = self else {
                return
            }
            self?.locationFetchTimer = Timer.scheduledTimer(timeInterval: fetchService.interval, target: fetchService, selector: #selector(fetchService.onTimerFire), userInfo: nil, repeats: false)
        }
    }
    
    @objc func onTimerFire() {
        if !UserUtility.singleton.isTokenExpired {
            pullLocations()
        }
    }
}
